source/partitions.c

```diff
@@ -0,0 +1,269 @@
+        partition_info = aws_partition_info_new(wrapper->partitions->allocator, *key);
+        partition_info->info = aws_string_new_from_json(wrapper->partitions->allocator, merge.merge_node);
+        aws_json_value_destroy(merge.merge_node);
+    } else {
+        partition_info = aws_partition_info_new(wrapper->partitions->allocator, *key);
+        partition_info->info = wrapper->outputs_str;
+        partition_info->is_copy = true;
+    }
+
+    if (aws_hash_table_put(
+            &wrapper->partitions->region_to_partition_info, &partition_info->name, partition_info, NULL)) {
+        AWS_LOGF_ERROR(AWS_LS_SDKUTILS_PARTITIONS_PARSING, "Failed to add partition info.");
+        goto on_error;
+    }
+
+    return AWS_OP_SUCCESS;
+
+on_error:
+    if (partition_info != NULL) {
+        aws_partition_info_destroy(partition_info);
+    }
+    return aws_raise_error(AWS_ERROR_SDKUTILS_PARTITIONS_PARSE_FAILED);
+}
+
+static int s_on_partition_element(
+    size_t idx,
+    const struct aws_json_value *partition_node,
+    bool *out_should_continue,
+    void *user_data) {
+    (void)out_should_continue;
+    (void)idx;
+
+    struct aws_partitions_config *partitions = user_data;
+
+    struct aws_byte_cursor id_cur;
+    struct aws_json_value *id_node = aws_json_value_get_from_object(partition_node, aws_byte_cursor_from_c_str("id"));
+    if (id_node == NULL || aws_json_value_get_string(id_node, &id_cur)) {
+        AWS_LOGF_ERROR(AWS_LS_SDKUTILS_PARTITIONS_PARSING, "Failed to extract id of partition.");
+        goto on_error;
+    }
+
+    struct aws_json_value *outputs_node =
+        aws_json_value_get_from_object(partition_node, aws_byte_cursor_from_c_str("outputs"));
+    if (outputs_node == NULL) {
+        AWS_LOGF_ERROR(AWS_LS_SDKUTILS_PARTITIONS_PARSING, "Failed to extract outputs of partition.");
+        goto on_error;
+    }
+
+    struct aws_partition_info *partition_info = aws_partition_info_new(partitions->allocator, id_cur);
+    partition_info->info = aws_string_new_from_json(partitions->allocator, outputs_node);
+
+    if (partition_info->info == NULL) {
+        AWS_LOGF_ERROR(AWS_LS_SDKUTILS_PARTITIONS_PARSING, "Failed to add partition info.");
+        goto on_error;
+    }
+
+    if (aws_hash_table_put(&partitions->region_to_partition_info, &partition_info->name, partition_info, NULL)) {
+        AWS_LOGF_ERROR(AWS_LS_SDKUTILS_PARTITIONS_PARSING, "Failed to add partition info.");
+        goto on_error;
+    }
+
+    struct partition_parse_wrapper wrapper = {
+        .outputs_node = outputs_node, .outputs_str = partition_info->info, .partitions = partitions};
+
+    struct aws_json_value *regions_node =
+        aws_json_value_get_from_object(partition_node, aws_byte_cursor_from_c_str("regions"));
+    if (regions_node != NULL && aws_json_const_iterate_object(regions_node, s_on_region_element, &wrapper)) {
+        AWS_LOGF_ERROR(AWS_LS_SDKUTILS_PARTITIONS_PARSING, "Failed to parse regions.");
+        goto on_error;
```

Comment:
this is a bit concern, on this failure, the `partitions->region_to_partition_info` has added the new `partition_info`. Can we move this before the table put? Or, we actually don't care?

Reply:
dont think its a concern. partitions file has a partition specific meta data section and then each region within partition can override parts of that metadata, which will create a new region specific entry. If region parsing fails, then we partition wide metadata will be cleaned up in the top level clean up